### PR TITLE
Windows compatibility

### PR DESCRIPTION
**Pull request by sbrugman**
CI testing for Windows to prevent windows-specific issues in the future (https://github.com/charliermarsh/ruff/issues/2026,  https://github.com/charliermarsh/ruff/issues/1915, https://github.com/charliermarsh/ruff/issues/832)

### CODE REVIEW

**Comment by charliermarsh** [comment]
I feel like I'm missing something hah - how does `cargo test` pass on Windows if there are failing tests?

**Comment by sbrugman** [comment]
Probably something with powershell not passing the exit code to github actions.

e.g. in this job https://github.com/charliermarsh/ruff/actions/runs/3969891762/jobs/6805182473:
```
failures:
[1018](https://github.com/charliermarsh/ruff/actions/runs/3969891762/jobs/6805182473#step:7:1019)
    source_code::generator::tests::indent
[1019](https://github.com/charliermarsh/ruff/actions/runs/3969891762/jobs/6805182473#step:7:1020)
    source_code::generator::tests::set_indent
```

**Comment by messense** [comment]
Maybe just add `shell: bash` to that `run` step.

**Comment by charliermarsh** [comment]
(In the interest of clear communication: I'd love to get this working!)

File: src/assert_yaml_snapshot.rs [hunk 1]
```diff
@@ -0,0 +1,15 @@
+#[macro_export]
+macro_rules! assert_yaml_snapshot {
+    ( $($args: expr),+) => {
+        if cfg!(windows){
+            insta::with_settings!({filters => vec![
+                (r"\\r\\n", "\\n"),
```
[the remark this answers is not in the document]
Comment (sbrugman) [reply src/assert_yaml_snapshot.rs:6]: See comment below

**Comment by sbrugman** [comment]
@charliermarsh Many files changed, but the actual changes are manageable (especially if you look per commit)

**Comment by charliermarsh** [comment]
@sbrugman - Thank you! Heroic work. I'll try to review today.

**Comment by charliermarsh** [comment]
Yeah will do. Don't want you to have to keep rebasing here. I'll review this tonight, and hopefully we can merge tomorrow or whenever's convenient.

**Comment by charliermarsh** [comment]
> Current errors on windows are caused by the new EXE features.

(Are those real failures?)

File: src/assert_yaml_snapshot.rs [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+macro_rules! assert_yaml_snapshot {
+    ( $($args: expr),+) => {
+        let line_sep = if cfg!(windows) { "\r\n" } else { "\n" };
+
+        // adjust snapshot file for platform
+        let mut settings = insta::Settings::clone_current();
+        settings.add_redaction("[].fix.content", insta::dynamic_redaction(move |value, _path| {
+            // CRLF/LF => <NEWLINE>
+            // TODO: do not replace escaped!
+            value.as_str().unwrap().replace(line_sep, "<NEWLINE>")
```
Comment (andersk) [inline src/assert_yaml_snapshot.rs:12]: For more readable output, we could _split_ on newlines instead, yielding a list.

```suggestion
            insta::internals::Content::Seq(
                value.as_str().unwrap().split(line_sep).map(|line| line.into()).collect()
            )
```

Comment (sbrugman) [reply src/assert_yaml_snapshot.rs]: Much better!

**Comment by charliermarsh** [comment]
(I'm reviewing this now.)

File: src/assert_yaml_snapshot.rs [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+    ( $($args: expr),+) => {
+        let line_sep = if cfg!(windows) { "\r\n" } else { "\n" };
+
+        // adjust snapshot file for platform
+        let mut settings = insta::Settings::clone_current();
+        settings.add_redaction("[].fix.content", insta::dynamic_redaction(move |value, _path| {
+            // CRLF/LF => <NEWLINE>
+            // TODO: do not replace escaped!
+            insta::internals::Content::Seq(
+                value.as_str().unwrap().split(line_sep).map(|line| line.into()).collect()
```
Comment (charliermarsh) [inline src/assert_yaml_snapshot.rs:13]: What would happen if we used `.lines()` here instead? Apart from losing the trailing newline / empty line.

File: src/rules/isort/mod.rs [hunk 4]
```diff
@@ -738,10 +737,29 @@ mod tests {
                 ..Settings::for_rule(Rule::UnsortedImports)
             },
         )?;
-        insta::assert_yaml_snapshot!(snapshot, diagnostics);
+        assert_yaml_snapshot!(snapshot, diagnostics);
         Ok(())
     }
 
+    // Test currently disabled as line endings are automatically converted to
+    // platform-appropriate ones in CI/CD #[test_case(Path::new("
```
Comment (charliermarsh) [inline src/rules/isort/mod.rs:745]: Nit: the code (`#` onwards) should be on the next line, I think.

**Comment by charliermarsh** [comment]
Looks great! Couple questions in the review. I have no doubt that this will catch a ton of issues.

**Comment by charliermarsh** [comment]
(I'll prioritize merging this as soon as the review is addressed, and try to avoid merging any conflicting changes in the interim!)

File: .github/workflows/ci.yaml [hunk 5]
```diff
@@ -63,9 +63,14 @@ jobs:
...
-    name: "cargo test"
-    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        os: [ ubuntu-latest, windows-latest ]
+    runs-on: ${{ matrix.os }}
+    name: "cargo test | ${{ matrix.os }}"
     steps:
+      # - name: Prepare Git
+      #   run: git config --global core.autocrlf input
```
[the remark this answers is not in the document]
Comment (sbrugman) [reply .github/workflows/ci.yaml:73]: Yes. I left it intentionally before to emphasise that git is automatically changing the line endings which, together with the line ending detection, is causing the isort test to fail (but without it many other tests do not work).

File: src/rules/isort/rules/add_required_imports.rs [hunk 6]
```diff
@@ -98,6 +98,7 @@ fn contains(block: &Block, required_import: &AnyImport) -> bool {
 }
 
 fn add_required_import(
+    stylist: &Stylist,
```
[the remark this answers is not in the document]
Comment (sbrugman) [reply src/rules/isort/rules/add_required_imports.rs:101]: ✅

File: src/rules/isort/mod.rs [hunk 7]
```diff
@@ -738,10 +737,29 @@ mod tests {
...
+    //             .as_path(),
+    //         &Settings {
+    //             src:
+    // vec![Path::new("resources/test/fixtures/isort").to_path_buf()],
+    //             ..Settings::for_rule(Rule::UnsortedImports)
+    //         },
+    //     )?;
+    //     insta::assert_yaml_snapshot!(snapshot, diagnostics);
+    //     Ok(())
+    // }
```
[the remark this answers is not in the document]
Comment (sbrugman) [reply src/rules/isort/mod.rs:761]: The test does pass locally. The issue is that when we checkout the repository, git reformats the line endings of the test files. This is expected and required for most tests, but in the case of `line_endings_crlf.py` and `line_endings_lf.py`, git should not touch them.

A workaround would be to not store these two files in a python file but as a testing string in the file.

**Comment by charliermarsh** [comment]
@sbrugman - What's the current state of this PR? Is it mergeable? I notice that you've been tweaking the EXE tests a bit, but it seems that the changes are now causing different CI failures.

**Comment by sbrugman** [comment]
@charliermarsh Something is going wrong with selecting the right snap files per platform, I believe. 
I don't have the time to work on it today. Perhaps we can ignore these tests on Windows for now (with `cfg`), then merge and then review it later

**Comment by charliermarsh** [comment]
Yeah that's fine. Do you mind if I debug a bit, push to this PR, then squash and merge?

**Comment by sbrugman** [comment]
> Yeah that's fine. Do you mind if I debug a bit, push to this PR, then squash and merge?

Perfect 🙏 !

**Comment by charliermarsh** [comment]
Okay thanks! I'll try to get this merged today then. And I'll look to preserve a few separate commits, but I may squash some of the intermediary commits.

**Comment by charliermarsh** [comment]
@sbrugman - I was thinking of squashing this down into a few commits, by squashing all the "core" and "fix" commits, so that we have that first commit, then all the "refactor" commits. Is that ok?

**Comment by sbrugman** [comment]
> @sbrugman - I was thinking of squashing this down into a few commits, by squashing all the "core" and "fix" commits, so that we have that first commit, then all the "refactor" commits. Is that ok?

Sounds good!

**Comment by charliermarsh** [comment]
(I'll merge this tonight, I'm just waiting for the v0.0.234 release to publish.)

**Comment by sbrugman** [comment]
@charliermarsh I could rebase and squash now if there are no more tweaks from your side (got 30 mins or so)

**Comment by charliermarsh** [comment]
Go for it!

**Comment by charliermarsh** [comment]
After this merges, I'll look into re-enabling those disabled tests.

**Comment by sbrugman** [comment]
@charliermarsh The 6 commits seem meaningful to me. The prefix `feat` is typically the only one that shows up in the changelog. The fixes grouped by issue may be helpful as reference if other contributors run into issues with the windows tests.

**Comment by charliermarsh** [comment]
Awesome. Thank you @sbrugman! This is great for Ruff, I know it was a lot of work.

**Pull request merged**